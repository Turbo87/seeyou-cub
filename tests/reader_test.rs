use seeyou_cub::{CubStyle, parse};
use std::fs::File;

#[test]
fn parse_france_fixture() {
    let file =
        File::open("tests/fixtures/france_2024.07.02.cub").expect("Failed to open fixture file");

    let (mut cub, warnings) = parse(file).expect("Failed to parse CUB file");

    // Validate warnings
    assert_eq!(warnings.len(), 0);

    // Snapshot the header
    insta::assert_debug_snapshot!("header", cub.header());

    // Validate item count
    assert_eq!(cub.items().len(), 1368);

    // Snapshot first few items
    insta::assert_debug_snapshot!("items_sample", &cub.items()[0..5]);

    // Validate first item's points
    let first_item = cub.items().first().unwrap().clone();
    let points: Vec<_> = cub
        .read_points(&first_item)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("first_item_points", points);

    let last_item = cub.items().last().unwrap().clone();
    let points: Vec<_> = cub
        .read_points(&last_item)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("last_item_points", points);

    // Validate all airspaces and collect statistics
    let mut total_points = 0;
    let mut style_counts = std::collections::HashMap::new();

    let items = cub.items().to_vec();
    for item in &items {
        *style_counts.entry(item.style()).or_insert(0) += 1;

        let mut point_iter = cub.read_points(item).expect("Failed to read points");

        for point_result in &mut point_iter {
            let _point = point_result.expect("Failed to parse point");
            total_points += 1;
        }
    }

    // Validate total points
    assert_eq!(total_points, 93784, "Total points mismatch");

    // Validate airspace type distribution
    assert_eq!(
        *style_counts.get(&CubStyle::Unknown).unwrap_or(&0),
        452,
        "Unknown airspace count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::RestrictedArea).unwrap_or(&0),
        435,
        "Restricted area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ProhibitedArea).unwrap_or(&0),
        113,
        "Prohibited area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::DangerArea).unwrap_or(&0),
        111,
        "Danger area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::GliderSector).unwrap_or(&0),
        135,
        "Glider sector count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ControlZone).unwrap_or(&0),
        92,
        "Control zone count mismatch"
    );
}
