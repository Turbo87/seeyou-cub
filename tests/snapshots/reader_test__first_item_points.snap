---
source: tests/reader_test.rs
expression: points
---
[
    ParsedPoint {
        lon: 0.033180647,
        lat: 0.83465517,
        name: None,
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
    ParsedPoint {
        lon: 0.033224326,
        lat: 0.8346504,
        name: None,
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
    ParsedPoint {
        lon: 0.03319982,
        lat: 0.83466476,
        name: None,
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
    ParsedPoint {
        lon: 0.033180647,
        lat: 0.83465517,
        name: None,
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
    ParsedPoint {
        lon: 0.0332435,
        lat: 0.8346504,
        name: Some(
            "USE\u{1}\u{1b}\0R\0\u{1}\0\0\u{1b}",
        ),
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
    ParsedPoint {
        lon: 0.033258412,
        lat: 0.83467966,
        name: None,
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
    ParsedPoint {
        lon: 0.03319503,
        lat: 0.8346941,
        name: None,
        frequency: None,
        frequency_name: None,
        optional_data: [],
    },
]
